src/events.ts

```diff
@@ -0,0 +1,86 @@
+export type TurboBeforeVisitEvent = CustomEvent<{ url: string }>
+export type TurboClickEvent = CustomEvent<{ url: string; originalEvent: MouseEvent }>
+export type TurboFrameLoadEvent = CustomEvent
+export type TurboBeforeFrameRenderEvent = CustomEvent<{ newFrame: FrameElement } & FrameViewRenderOptions>
+export type TurboFrameRenderEvent = CustomEvent<{ fetchResponse: FetchResponse }>
+export type TurboLoadEvent = CustomEvent<{ url: string; timing: TimingData }>
+export type TurboRenderEvent = CustomEvent
+export type TurboReloadEvent = CustomEvent
+export type TurboVisitEvent = CustomEvent<{ url: string; action: Action }>
+
+export type TurboBeforeStreamRenderEvent = CustomEvent<{ newStream: StreamElement; render: Render }>
+
+export type TurboSubmitStartEvent = CustomEvent<{ formSubmission: FormSubmission }>
+export type TurboSubmitEndEvent = CustomEvent<
+  { formSubmission: FormSubmission } & { [K in keyof FormSubmissionResult]?: FormSubmissionResult[K] }
+>
+
+export type TurboFrameMissingEvent = CustomEvent<{ response: Response; visit: VisitFallback }>
+
+export type TurboBeforeFetchRequestEvent = CustomEvent<{
+  fetchOptions: RequestInit
+  url: URL
+  resume: (value: any) => void
+}>
+
+export type TurboBeforeFetchResponseEvent = CustomEvent<{
+  fetchResponse: FetchResponse
+}>
+
+export type TurboFetchRequestErrorEvent = CustomEvent<{
+  request: FetchRequest
+  error: Error
+}>
+
+interface TurboGlobalEventMap {
+  "turbo:before-cache": TurboBeforeCacheEvent
+  "turbo:before-fetch-request": TurboBeforeFetchRequestEvent
+  "turbo:before-fetch-response": TurboBeforeFetchResponseEvent
+  "turbo:before-frame-render": TurboBeforeFrameRenderEvent
+  "turbo:before-stream-render": TurboBeforeStreamRenderEvent
+  "turbo:before-render": TurboBeforeRenderEvent
+  "turbo:before-visit": TurboBeforeVisitEvent
+  "turbo:click": TurboClickEvent
+  "turbo:fetch-request-error": TurboFetchRequestErrorEvent
+  "turbo:frame-load": TurboFrameLoadEvent
+  "turbo:frame-render": TurboFrameRenderEvent
+  "turbo:frame-missing": TurboFrameMissingEvent
+  "turbo:load": TurboLoadEvent
+  "turbo:render": TurboRenderEvent
+  "turbo:reload": TurboReloadEvent
+  "turbo:submit-start": TurboSubmitStartEvent
+  "turbo:submit-end": TurboSubmitEndEvent
+  "turbo:visit": TurboVisitEvent
+}
+
+interface TurboElementEventMap {
+  "turbo:before-frame-render": TurboBeforeFrameRenderEvent
+  "turbo:before-fetch-request": TurboBeforeFetchRequestEvent
+  "turbo:before-fetch-response": TurboBeforeFetchResponseEvent
+  "turbo:fetch-request-error": TurboFetchRequestErrorEvent
+  "turbo:frame-load": TurboFrameLoadEvent
+  "turbo:frame-render": TurboFrameRenderEvent
+  "turbo:frame-missing": TurboFrameMissingEvent
+  "turbo:submit-start": TurboSubmitStartEvent
+  "turbo:submit-end": TurboSubmitEndEvent
+}
+
+declare global {
+  interface DocumentEventMap extends TurboGlobalEventMap {}
+  interface WindowEventMap extends TurboGlobalEventMap {}
```

Comment:
@marcoroth does this clobber the existing `WindowEventMap` interface? For example, if I were to add a `"click"` event listener, would it be able to infer the type as `MouseEvent`?

Reply:
That should still work since this is just extendeding the existing interface

Reply:
It might be due to misconfigurations in my editor (Vim), but when I try to jump to the definition for `WindowEventMap`, it brings me to this one, instead of the TypeScript built-in one with all of the event names.

Is there a way to hint to the code to disambiguate between extending an existing `interface` and declaring a new one?

Reply:
I'm not sure if there's a different way of doing this. But my gut tells me it's more of an editor-tooling/LSP thing.